torch/csrc/jit/passes/onnx/shape_inference.cpp

```diff
@@ -0,0 +1,168 @@
+  }
+
+  if (node_kind == ::c10::onnx::SequenceAt ||
+      node_kind == ::c10::onnx::SplitToSequence ||
+      node_kind == ::c10::onnx::SequenceConstruct ||
+      node_kind == ::c10::onnx::SequenceEmpty ||
+      node_kind == ::c10::onnx::SequenceInsert ||
+      node_kind == ::c10::onnx::ConcatFromSequence ||
+      node_kind == ::c10::onnx::SequenceErase) {
+    // TODO: ONNX unable to do shape inference for these ops.
+    return false;
+  }
+
+  if (node_kind == ::c10::onnx::ConstantOfShape) {
+    // && n->input()->node()->kind() == ::c10::prim::ListConstruct
+    // TODO: ONNX shape inference segfault.
+    return false;
+  }
+
+  if (node_kind == ::c10::onnx::Loop || node_kind == ::c10::onnx::If) {
+    // TODO: Support Loop & If shape inference by propagating input shape to
+    // block input.
+    return false;
+  }
+
+  return true;
+}
+
+} // namespace
+
+void ONNXShapeInference(Node* n, int opset_version) {
+  if (!IsSupportedNode(n)) {
+    return;
+  }
+
+  auto node_kind = n->kind();
+
+  GRAPH_UPDATE("Shape inference for node kind: ", node_kind.toDisplayString());
+  auto g = n->owningGraph();
+
+  auto node_g = std::make_shared<Graph>();
+  auto clone_node = node_g->createClone(n, [&node_g](Value* v) {
+    auto v_n = v->node();
+    if (v_n->kind() == ::c10::onnx::Constant) {
+      auto constant_n = node_g->insertNode(
+          node_g->createClone(v_n, [](Value* v) { return v; }));
+      return constant_n->output();
+    } else {
+      auto input = node_g->addInput();
+      input->copyMetadata(v);
+      return input;
+    }
+  });
+  node_g->insertNode(clone_node);
+  for (auto output : clone_node->outputs()) {
+    node_g->registerOutput(output);
+  }
+  GRAPH_DUMP("New temporary graph:", node_g);
+
+  // TODO: Some ops have conversion happen at Peephole pass.
+  //       The conversion here is incomplete for these ops.
+  //       e.g: ListConstruct, ListUnpack, etc.
+  std::string model_str;
+  RawDataExportMap export_map;
+  std::tie(model_str, export_map) = export_onnx(
+      node_g,
+      {},
+      opset_version,
+      {},
+      false,
+      onnx_torch::OperatorExportTypes::ONNX,
+      true,
+      true,
+      {},
+      true,
+      false,
+      std::string());
+  onnx::ModelProto model_proto;
+  model_proto.ParseFromString(model_str);
+  GRAPH_UPDATE("generated proto ", prettyPrint(model_proto));
+
+  // infer shape
+  onnx::shape_inference::InferShapes(model_proto);
+  GRAPH_UPDATE("generated proto with shape ", prettyPrint(model_proto));
+
+  auto graph_proto = model_proto.graph();
+
+  // inferred shapes are stored in value_info.
+  for (size_t i = 0; i < graph_proto.value_info_size(); ++i) {
+    auto v_info = graph_proto.value_info(i);
+    // get data from value_info and updated original graph.
+    for (size_t j = 0; j < clone_node->outputs().size(); ++j) {
+      if (clone_node->output(j)->debugName() == v_info.name()) {
+        UpdateValueByValueInfo(n->output(j), v_info);
```

Comment:
This method name `UpdateValueByValueInfo` gives me a pause to understand what it does. Would it be better to give it a more explicit name such as `UpdateTorchValueByOnnxValueInfo`?